### PR TITLE
Bubble sort by feature

### PR DESCRIPTION
**Pull request by george-shammar**
This Pull Request seeks to merge the content of bubble_sort_by_feature to the main branch.

Below are the descriptions of the feature updates:

1. The feature contains a defined method, bubble_sort_by based on the bubble_sort model.
2. The feature is meant to be called on an array.
3. The algorithm compare two items of an array - arr[i] and arr[i + 1] or left and right as defined by the left and right variable in the code block.
4. The code block is meant to be executed arr.length.times, representing the number of items in the array.
5. If arr[i] is greater than arr[i + 1], the two items' positions are to be swapped.
6. The new method, bubble_sort_by was called on a test array and the array items were sorted.

### CODE REVIEW

**Review by ebeagusamuel — CHANGES_REQUESTED** [review]
# Changes Requested :recycle: 
### Hey there :wave: 
This is a nice first attempt at this project. However, there are still some changes to be made.
- [ ] Github action is failing which means you have some rubocop error. Please do fix them as the project cannot be approved with these errors.
- [x] Please refer back to the project description. You are to create two methods, i.e bubble_sort, and bubble_sort_by. You only have the bubble_sort_by method.
- [ ] Also remember that the bubble_sort_by method should be able to accept a block, but it doesn't currently. Please refer back to the project description and make changes accordingly.
#### Request another review when you are done making the changes.
slack(@ebeagusamuel)

File: bubble_sort.rb [hunk 1]
```diff
@@ -0,0 +1,29 @@
...
+  arr.length.times do
+    arr.each_with_index do |_x, i|
+      left = arr[i]
+      if arr[-1] != left
+        arr[i], arr[i + 1] = arr[i + 1], arr[i] if (arr[i].length - arr[i + 1].length).positive?
+      end
+    end
+  end
+  arr
+end
```
Comment (voscarmv) [inline bubble_sort.rb:27]: The `bubble_sort_by` method is supposed to take in a code block as an argument. The purpose of this code block is to provide the programmer with a way of specifying the type of sorting they want to execute. For example, the programmer might want to implement a reverse sort, or they may want to do a lexicographic sort.

An example of how the `bubble_sort_by` method sould be used would be:

```
unsorted = %w[hello hiya hey hi h]
# This should sort by ascending length:
bubble_sort_by(unsorted) { |left, right| left.length - right.length }

# Whereas this should sort by descending length
bubble_sort_by(unsorted) { |left, right| right.length - left.length }
```

The idea is to write a **flexible** method that allows the programmer to specify the type of sorting via the code block. As The Odin Project states, for this you may use the `yield` method provided by ruby:

image

**Review by voscarmv — CHANGES_REQUESTED** [review]
# Nice work so far. 👍 

## However, there are some changes required 👀 🧰 🔨 🔧 

- Please read my code comments for the required changes.

### Submit these changes and request a new code review! 😄 👍 

### Questions or comments? Reach me on [Twitter](https://twitter.com/voscarmv) or [LinkedIn](https://linkedin.com/in/oscar-mier) 🧑‍💻🗨️🧑‍💻 
# Happy Hacking! 🧠 ⌨️ 🖱️ 🖥️

**Review by jubaan — CHANGES_REQUESTED** [review]
# Status: Changes Required :recycle: 

There are some issues that need to be addressed before we can move forward. Please review the list below.

## Required changes :wrench: 
### [Mandatory] :red_circle: 

- [x] The `bubble_sort_by` is not sorting the elements in the correct order. 

```ruby
expected: ["h", "hi", "hey", "hiya", "hello"]
     got: ["hello", "hiya", "hey", "hi", "h"]
```

Please fix this and submit it again for a re-review.

## Keep up the great work :muscle: and Happy Coding :smiley: :computer:!!

File: bubble_sort.rb [hunk 2]
```diff
@@ -0,0 +1,31 @@
...
+
+puts bubble_sort([4, 3, 78, 2, 0, 2])
+
+# bubble_sort_by
+def bubble_sort_by(arr)
+  arr.length.times do
+    arr.each_with_index do |_x, i|
+      left = arr[i]
+      if arr[-1] != left
+        yield arr[i], arr[i + 1] = arr[i + 1], arr[i] if (arr[i].length - arr[i + 1].length).positive?
```
Comment (voscarmv) [inline bubble_sort.rb:22]: In line 21 you compare `arr[-1]` with `left`. But note that `arr[-1]` will represent the last element of the array, and not the `i-1`th element (which is what you probably meant). Also, notice that the `i-1` element would be to the *left* of the `i`th element in the array. Please fix this.

Also, in line 22, you use `yield` incorrectly. The `yield` method is supposed to be used to determine the output of the block being used. Please see the notes below on how you have also misused the block in this file.

File: bubble_sort.rb [hunk 3]
```diff
@@ -0,0 +1,31 @@
...
+      end
+    end
+  end
+  arr
+end
+
+p bubble_sort_by(%w[hi hello h hiya hey]) do
+  left.length - right.length
+end
```
Comment (voscarmv) [inline bubble_sort.rb:31]: In this part of the code, you are referring to `left` and `right` even though you never really defined them anywhere in the code. The correct way to call `bubbble_sort_by` here would be:

```
p bubble_sort_by(%w[hi hello h hiya hey]) { |left, right|
  left.length - right.length
}
```

**Review by voscarmv — CHANGES_REQUESTED** [review]
# Impressive work, team. 💻 👀 ✨ 

## However, there are some changes required 👀 🧰 🔨 🔧 

- There are several problems with your `bubble_sort_by` method. Please read my code comments for hints on how to fix them.

### Submit these changes and request a new code review! 😄 👍 

### Questions or comments? Reach me on [Twitter](https://twitter.com/voscarmv) or [LinkedIn](https://linkedin.com/in/oscar-mier) 🧑‍💻🗨️🧑‍💻 
# Happy Hacking! 🧠 ⌨️ 🖱️ 🖥️

**Comment by voscarmv** [comment]
You may refer to this article for more information on how to use the `yield` method in ruby with multiple input parameters (as required by this project)

https://stackoverflow.com/questions/41865139/ruby-block-taking-array-or-multiple-parameters

**Review by voscarmv — CHANGES_REQUESTED** [review]
# Nice work so far. 👍 
## However, there are some changes required 👀 🧰 🔨 🔧 


- Your code is still not sorting the words in the array by length, which is what's expected:

image

The correct order here should be:

```
["h", "hi", "hey", "hiya", "hello"]
```

That is: the words are supposed to be sorted by length because that is the comparison that is being done in the code block:

```
p bubble_sort_by(%w[hi hello h hiya hey]) { |left, right|
  left.length - right.length # <-- Here
}
```

It seems to me that you haven't quite grasped the use of code blocks and `yeld` in ruby. Don't worry :) Please study the following two articles to gain a better understanding of these subjects:

> - https://mixandgo.com/learn/ruby-blocks
> - https://stackoverflow.com/questions/36959934/how-can-i-pass-in-a-block-to-my-bubble-sort-method

If you still need help, I'll be available on slack in a few hours. Drop me a message if you want to schedule a coding session so I can help you out with this.

Don't feel discouraged, you're almost there! 

### Submit these changes and request a new code review! 😄 👍 

### Questions or comments? Reach me on [Twitter](https://twitter.com/voscarmv) or [LinkedIn](https://linkedin.com/in/oscar-mier) 🧑‍💻🗨️🧑‍💻 
# Happy Hacking! 🧠 ⌨️ 🖱️ 🖥️

**Review by voscarmv — APPROVED** [review]
# Project status: Approved ✨ 😎👍 💯 ✔️

- Excellent work catching on to the hints in the articles. Well done!

### You may merge this branch and move forward ⛙🛣

### Questions or comments? Reach me on [Twitter](https://twitter.com/voscarmv) or [LinkedIn](https://linkedin.com/in/oscar-mier) 🧑‍💻🗨️🧑‍💻 
# Happy Hacking! 🧠 ⌨️ 🖱️ 🖥️

**Pull request merged**